text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+
+### Proposal response consistency checking
+
+Current client SDK implementations do minimal checking of proposal responses before sending to the orderer. There have been some requests for more rigorous checking of proposal responses prior to submit to the orderer.
+
+Prior to sending to the orderer, Fabric Gateway should check that proposal responses are:
+- Consistent, including having matching response payloads.
+- Meet all applicable endorsement requirements.
+
+### Ensure query results reflect latest ledger state
+
+Query results should be obtained from peers whose ledger height is at least that of the most recent events sent to clients.
+
+Scenario:
+- Client receipt of chaincode event triggers business process.
+- Business process queries ledger to obtain private data, which cannot be included in event payload since it is private.
+The private data must be consistent with the state of the ledger after commit of the transaction that emitted the chaincode event, so the query must be processed by a peer with ledger height of at least the block number that contained the chaincode event.
+
+Implementation options:
+- Only direct queries (and endorsement requests) to the peer(s) with the highest ledger height.
+- Client proposals include last observed ledger height so Gateway can target any peers with at least that ledger height.
+ A view of ledger height can be tracked by the Gateway from block events it has received.
+
+## Scaling and load balancing
+
+Currently a client application is responsible for load balancing its requests between multiple peers.  The Gateway will handle this on behalf of the set of currently connected client applications.
+
+Ideally, the Gateway itself will be stateless (i.e. not maintain client session state) allowing clients to be routed though an appriopriate load balancer.
+
+## Gateway SDKs
+
+A set of SDKs will be created to allow client applications to interact with the Fabric Gateway.  The APIs exposed by these SDKs will be, where possible, the same as the current high-level 'gateway' SDKs.
+
+The following classes/structures will be available as part of the gateway programming model:
+- Gateway
+- Network
+- Contract
+- Transaction
+- Wallet
+
+The existing `submitTransaction()` and `evaluateTransaction()` API methods will be implemented using the Gateway gRPC services described above.
+
+Based on user feedback of the existing gateway SDKs, the following additions are planned to be supported by this implementation:
+- Support for 'offline signing', allowing clients to write REST servers that don't directly have access to users' private keys.
+- Support for 'asynchronous submit', allowing client applications to continue processing while the transaction commit proceeds in the background. For Go this will be implemented by returning a channel; for Node by returning a Promise; and for Java by returning a Future.
```

Comment:
Going back to my earlier comment, this means the asynchronous nature is intended to be part of the client and not the gateway service. Correct?

Unless we have a really good reason for returning streams from gRPC, it's best to avoid them.

Reply:
https://github.com/hyperledger/fabric-rfcs/pull/32#discussion_r520477985